### PR TITLE
feat(applicationset): Add FlatList option to cluster generator - Fixes #20212

### LINKED ISSUE #20212: ApplicationSet: accessing cluster list inside template without creating a single element per cluster
# Summary

If we want to access a list of cluster used inside ArgoCD, we generally use the clusterGenerator

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: guestbook
  namespace: argocd
spec:
  goTemplate: true
  goTemplateOptions: ["missingkey=error"]
  generators:
  - clusters:       
      selector:
        matchLabels:
          staging: "true"
  template:
    metadata:
      name: '{{.name}}-guestbook' # 'name' field of the Secret
    spec:
      project: "my-project"
      source:
        repoURL: https://github.com/argoproj/argocd-example-apps/
        targetRevision: HEAD
        path: guestbook
      destination:
        server: '{{.server}}' # 'server' field of the secret
        namespace: guestbook
```
However, this ends up creating one Application per cluster selected.
It is useful to create one Application in each of those clusters, but sometimes we only need to get the clusters info without using them as destination.

Would there be a way to get the list of cluster matched, and use them as a value list inside a single Application ?

Something like:
```yaml
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
metadata:
  name: guestbook
  namespace: argocd
spec:
  goTemplate: true
  goTemplateOptions: ["missingkey=error"]
  generators:
  - clusters:
      flatClusterMode: true #<------ Dummy option to use it flat
      selector:
        matchLabels:
          staging: "true"
  template:
    metadata:
      name: 'my-guestbook' 
    spec:
      project: "my-project"
      source:
        repoURL: https://github.com/argoproj/argocd-example-apps/
        targetRevision: HEAD
        path: guestbook
        helm:
          values: |
            clusters:
            {{ range .clusters }} #<--- access the cluster list
              - name: {{ .name }}
            {{ end }}
      destination:
        server: 'in-cluster' # <--- deploying it a single place
        namespace: guestbook
```

### PR DESCRIPTION
**Pull request by OpenGuidou**
Fixes  #20212

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [x] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [x] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: applicationset/generators/cluster.go [hunk 1]
```diff
@@ -149,11 +158,19 @@ func (g *ClusterGenerator) GenerateParams(appSetGenerator *argoappsetv1alpha1.Ap
...
+			clustersParams = append(clustersParams, params)
+		} else {
+			res = append(res, params)
+		}
 
 		log.WithField("cluster", cluster.Name).Info("matched cluster secret")
 	}
 
+	if isFlatMode {
+		clusters["clusters"] = clustersParams
```
Comment (speedfl) [inline applicationset/generators/cluster.go:171]: no need to instantiate clusters above:

```golang
res = append(res, map[string]interface{
  "clusters": clustersParams
})
```

**Comment by speedfl** [comment]
LGTM !

**Comment by speedfl** [comment]
@OpenGuidou Maybe the array will be overridden if we have multiple cluster generator. Could you please test it ?

**Comment by OpenGuidou** [comment]
> @OpenGuidou Maybe the array will be overridden if we have multiple cluster generator. Could you please test it ?

As the values and template can be different in each cluster generator, I would not risk to flat list the flat lists :smile: 
I would say, if you agree, that it's acceptable to keep the factorization to a single cluster generator only.

**Comment by speedfl** [comment]
> > @OpenGuidou Maybe the array will be overridden if we have multiple cluster generator. Could you please test it ?
> 
> As the values and template can be different in each cluster generator, I would not risk to flat list the flat lists 😄 I would say, if you agree, that it's acceptable to keep the factorization to a single cluster generator only.

I think we can accept that. A small comment in the doc could do the job

**Comment by speedfl** [comment]
@ishitasequeira if you have some time 🙂

File: docs/operator-manual/applicationset/Generators-Cluster.md [hunk 2]
```diff
@@ -252,3 +252,63 @@ spec:
         server: '{{.values.clusterName}}'
         namespace: guestbook
 ```
+### Gather cluster information as a flat list ###
```
Comment (crenshaw-dev) [inline docs/operator-manual/applicationset/Generators-Cluster.md:255]:
```suggestion
### Gather cluster information as a flat list
```

The trailing characters get rendered in the title in ReadTheDocs.

**Review by crenshaw-dev — APPROVED** [review]

**Pull request merged**